joplin/base/wagtail_hooks.py

```diff
@@ -300,6 +301,158 @@ def register_link_handler(features):
 
 
+# In here we're going to remake all the urls for this page
+@hooks.register('after_edit_page')
+def after_edit_page(request, page):
+    # Make sure we're publishing so our Janis urls reflect the live site
+    # taking this from https://github.com/wagtail/wagtail/blob/349fca66f32ecf3df6316b0cffb4e5fd0d24d6c7/wagtail/admin/views/pages.py#L228
+    publishing = bool(request.POST.get('action-publish'))
+    if not publishing:
+        return
+
+    # Clear out all old urls
+    for janis_url in page.janis_urls.all():
+        janis_url.janis_url.delete()
+
+    # Create the new urls
+    new_urls = []
+
+    # Page type matters here
+    page_type = page._meta.object_name
+
+    # todo: not have magic language arrays hiding around the codebase
+    languages = ['en', 'es']
+    for language in languages:
+        # If we're a topic collection page we only have one url
+        # /theme_slug/topic_collection_slug/
+        if page_type == 'TopicCollectionPage':
+            new_url = JanisUrl.create(
+                        topic_collection_page=page,
+                        theme=page.theme,
+                        page_type=page_type,
+                        language=language)
+            new_url.save()
+            new_urls.append(new_url)
+
+        # If we're a topic page, we have a url for ever topic collection we belong to
+        # /theme_slug/topic_collection_slug/topic_slug/
+        elif page_type == 'TopicPage':
+            for topic_page_topic_collection in page.topiccollections.all():
+                new_url = JanisUrl.create(
+                            topic_page=page,
+                            topic_collection_page=topic_page_topic_collection.topiccollection,
+                            theme=topic_page_topic_collection.topiccollection.theme,
+                            page_type=page_type,
+                            language=language)
+                new_url.save()
+                new_urls.append(new_url)
+
+        # If we're a department page we only have one url
+        # /department_slug/
+        elif page_type == 'DepartmentPage':
+            new_url = JanisUrl.create(
+                        department_page=page,
+                        page_type=page_type,
+                        language=language)
+            new_url.save()
+            new_urls.append(new_url)
+
+        # If we're a location page we only have one url
+        # /location/location_slug/
+        elif page_type == 'LocationPage':
+            new_url = JanisUrl.create(
+                        location_page=page,
+                        page_type=page_type,
+                        language=language)
+            new_url.save()
+            new_urls.append(new_url)
+
+        # If we're any other page type, we need to make all our urls
+        # /theme_slug/topic_collection_slug/topic_slug/page_slug/
+        # /department_slug/page_slug/
+        # /page_slug/
+        else:
+            for page_topic in page.topics.all():
```

Comment:
these two querysets (looping one inside another) are going to be slow

Reply:
How slow are you thinking? Pages generally have one topic, each topic generally has 1 topic collection, so I wasn't really worried about it.

Any suggestions for an alternative method?